### PR TITLE
ORC-422 - Fix issue with Predicate push down when lower/upper bounds are set

### PR DESCRIPTION
**Pull request by moresandeep**
This PR fixes a bug where Predicate push down will not return proper values when upper and lower bounds are set for StringStatisticsImpl.
This patch also contains removing some unused imports as part of cleanup.

### CODE REVIEW

File: java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java [hunk 1]
```diff
@@ -35,13 +35,8 @@
 import org.apache.orc.TimestampColumnStatistics;
 import org.apache.orc.TypeDescription;
 
-import java.nio.ByteBuffer;
-import java.nio.CharBuffer;
-import java.nio.charset.Charset;
-import java.nio.charset.StandardCharsets;
 import java.sql.Date;
 import java.sql.Timestamp;
-import java.util.Arrays;
```
Comment (dongjoon-hyun) [inline java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java]: Shall we exclude these unrelated changes? It makes backports difficult sometimes.

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java [hunk 2]
```diff
@@ -37,7 +37,6 @@
 import java.sql.Timestamp;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
-import java.util.Arrays;
```
Comment (dongjoon-hyun) [inline java/core/src/test/org/apache/orc/TestColumnStatistics.java]: ditto.

File: java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java (hunk 1 above)
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java]: Good point, but do we need to carry on with un-used imports for upcoming releases, they can accumulate over time. Also, for backporting it could be  relatively easy to add these dependencies (I am assuming manual backports). Let me know if you feel strongly about it.

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java (hunk 2 above)
Comment (moresandeep) [reply java/core/src/test/org/apache/orc/TestColumnStatistics.java]: Same as above.

File: java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java (hunk 1 above)
Comment (dongjoon-hyun) [reply java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java]: We had better split that kind of cosmetic commits and critical bug fix commits.

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java (hunk 2 above)
Comment (dongjoon-hyun) [reply java/core/src/test/org/apache/orc/TestColumnStatistics.java]: Usually, it's better to clean up this if you change the actual code in this file.

File: java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java (hunk 1 above)
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/ColumnStatisticsImpl.java]: @dongjoon-hyun I reverted these changes.

File: java/core/src/test/org/apache/orc/TestColumnStatistics.java (hunk 2 above)
Comment (moresandeep) [reply java/core/src/test/org/apache/orc/TestColumnStatistics.java]: @dongjoon-hyun, the changes are undone.

**Comment by moresandeep** [comment]
Hello @omalley I am updating a work in progress patch for you to look at. I have a bunch of lines commented in there which I'll take out in the final patch and cleanup the code, apologies for the rough code, I don't want to delay this patch further. 

Basically, the crux of the patch is in file RecordReaderImpl line # 325 and # 335, what I am doing here is depending on whether upperbound or lowerbound is set for corner cases where the predicate literal equals upperbound or lowerbound I am picking the appropriate value. 
e.g. 
```
/* since min value is truncated when we have compare=0, it means the predicate string is BEFORE the min value*/
    else if (minCompare == 0 && isLowerBoundSet) {
      return Location.BEFORE;
    }
```
and 
   ```
 /* if upperbound is set then location here will be AFTER */
    else if (maxCompare == 0 && isUpperBoundSet) {
      return Location.AFTER;
    }
```
I also, tried to write up UnitTests to test all the corner case scenarios that I could think. I could not find a way test greater than since there is no Operator 'GREATER_THAN', code suggest using 'NOT' Operator, is there an example that I can look at for something like NOT LESS_THAN.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 3]
```diff
@@ -318,16 +315,25 @@ public long getNext() {
    * @param <T> the type of the comparision
    * @return the location of the point
    */
-  static <T> Location compareToRange(Comparable<T> point, T min, T max) {
+  static <T> Location compareToRange(Comparable<T> point, T min, T max, boolean isLowerBoundSet, boolean isUpperBoundSet) {
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: I suggest that we instead do:
```static <T> Location compareToRange(Comparable<T> point, T min, T max, T lowerBound, T upperBound)```
so the normal case would look like:
```compareToRange(min, max, min, max);```
or in the more general case:
```compareToRange(stats.getMin(), stats.getMax(), stats.getLowerBound(), stat.getUpperBound());```

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 4]
```diff
@@ -341,7 +347,7 @@ public long getNext() {
    * @return the object for the maximum value or null if there isn't one
    */
   static Object getMax(ColumnStatistics index) {
-    return getMax(index, false);
+    return getMax(index, false, null);
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: This seems suspicious.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 5]
```diff
@@ -577,19 +645,21 @@ private static boolean shouldEvaluateBloomFilter(PredicateLeaf predicate,
   private static TruthValue evaluatePredicateMinMax(PredicateLeaf predicate, Object predObj,
       Object minValue,
       Object maxValue,
-      boolean hasNull) {
+      boolean hasNull,
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: I think you should make the same change here.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java [hunk 6]
```diff
@@ -577,19 +645,21 @@ private static boolean shouldEvaluateBloomFilter(PredicateLeaf predicate,
...
+      boolean isLowerBoundSet,
+      boolean isUpperBoundSet) {
     Location loc;
 
     switch (predicate.getOperator()) {
       case NULL_SAFE_EQUALS:
-        loc = compareToRange((Comparable) predObj, minValue, maxValue);
+        loc = compareToRange((Comparable) predObj, minValue, maxValue, isLowerBoundSet, isUpperBoundSet);
```
Comment (omalley) [inline java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: You also need to deal with the IN sarg.

Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: This should be fixed in the new commit.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java (hunk 5 above)
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: Should be fixed in the new commit.

File: java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java (hunk 3 above)
Comment (moresandeep) [reply java/core/src/java/org/apache/orc/impl/RecordReaderImpl.java]: Sure, fixed this.

**Comment by moresandeep** [comment]
Hello @omalley can you take a look at the new changes which addresses the fixes.

**Comment by omalley** [comment]
I just committed this. Thanks, Sandeep!
